### PR TITLE
[browser][non-icu] `HybridGlobalization` SortKey

### PR DESCRIPTION
**Pull request by ilonatommy**
Implements a chunk of web-api based globalization. Is a part of HybridGlobalization feature and contributes to https://github.com/dotnet/runtime/issues/79989.

Old, icu-based private API: `GlobalizationNative_GetSortKey`

New, non-icu private API: none, Web API does not expose SortKey. We throw PlatformNotSupportedException.

Affected public API (see: tests in CompareInfoTests.cs):

- CompareInfo.GetSortKey
- CompareInfo.GetSortKeyLength
- CompareInfo.GetHashCode

All changes in behavior are listed in docs\design\features\hybrid-globalization.md

This PR also refactors CompareInfo tests - it moves common properties/fields to `CompareInfoTestsBase`.

cc @SamMonoRT

### CODE REVIEW

**Comment by pavelsavara** [comment]
What are the size wins ? 
What are the use-cases which would be affected by this change ?

**Comment by ilonatommy** [comment]
> What are the size wins ? What are the use-cases which would be affected by this change ?

Size win:
lack of SortKey is a side effect of removing collations data from ICU. Overall size decrease on full WASM ICU file (`icudt.dat`) ~40% (329770 bytes -> 202330 bytes on brotli compressed).

Use-cases (only of changes from this PR, not the whole collation):
- CompareInfo.GetSortKey
- CompareInfo.GetSortKeyLength
- CompareInfo.GetHashCode

**Comment by ilonatommy** [comment]
> I mean what are the use-cases where those APIs are useful ?
> 
> This is what I tried https://github.com/search?q=CompareInfo.GetSortKey+lang%3AC%23+&type=code&p=3 I see mostly database oriented libraries.

Sorry, I did not get it. It seems you are right, mostly DBs.

> Would it be better if we provide bitwise/invariant sortkey instead of PNSE ? Just an idea.

It is not a bad idea but I think if we make it a rule for all locales, it could be confusing. User that calls e.g. `new CultureInfo("de-DE").CompareInfo.GetSortKey()` expects locale-sensitive sort key.
If they need invariant SortKey, they can switch on `InvariantGlobalization` and when they need locale-sensitive, functional SortKey, they should not use `HybridGlobalization`.
I can think about a use-case when your solution would be perfectly useful:
- the user needs to cut down on runtime size (`HybridGlobalization` on)
- they use sort key for locale-insensitive case, e.g.:  `CultureInfo.InvariantCulture.CompareInfo.GetSortKey()`
- but do not want to switch on `InvariantGlobalization` becuse they need locale-sensitive operations somewhere else.

I will stash this idea in the tracking issue in case there is a demand on the above case.

File: src/libraries/System.Private.CoreLib/src/System/Globalization/CompareInfo.cs [hunk 1]
```diff
@@ -1599,5 +1615,9 @@ public SortVersion Version
         }
 
         public int LCID => CultureInfo.GetCultureInfo(Name).LCID;
+
+#if TARGET_BROWSER
+        private static string GetPNSEText(string funcName) =>  $"{funcName} is not supported when HybridGlobalization=true. Disable it to load larger ICU bundle, then use this option.";
```
Comment (pavelsavara) [inline src/libraries/System.Private.CoreLib/src/System/Globalization/CompareInfo.cs:1620]: Should this be in resource file and localized ?

**Review by pavelsavara — APPROVED** [review]
LGMT

What is our plan for Android/iOS where maybe Hybrid could support these methods ?

**Comment by ilonatommy** [comment]
> What is our plan for Android/iOS where maybe Hybrid could support these methods ?

Android might have an invariant equivalent, it's dedicated for LDAP sorting: https://docs.oracle.com/javase/7/docs/api/javax/naming/ldap/SortKey.html. I do not see any culture-related SortKey.
For iOS I cannot find an equivalent, @mkhamoyan do you know about any?
Even if Hybrid SortKey will be supported on both of them, then PNSE will be left only for WASM.

File: src/libraries/System.Private.CoreLib/src/System/Globalization/CompareInfo.cs (hunk 1 above)
Comment (ilonatommy) [reply src/libraries/System.Private.CoreLib/src/System/Globalization/CompareInfo.cs:1620]: Good idea, I can see all exception text in PrivateCoreLib are stored like this, so you are right.

**Pull request merged**